ambry-network/src/main/java/com/github/ambry/network/http2/Http2PeerCertificateValidator.java

```diff
@@ -0,0 +1,117 @@
+import java.security.cert.X509Certificate;
+import java.util.Collection;
+import java.util.List;
+import java.util.regex.Pattern;
+import javax.net.ssl.SSLSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handler that validates the peer certificate on the new HTTP2 connection.
+ */
+@ChannelHandler.Sharable
+public class Http2PeerCertificateValidator extends ChannelInboundHandlerAdapter {
+  private static final Logger logger = LoggerFactory.getLogger(Http2PeerCertificateValidator.class);
+  private final String http2PeerCertificateSanRegex;
+  private final Http2ClientMetrics http2ClientMetrics;
+
+  public Http2PeerCertificateValidator(String http2PeerCertificateSanRegex,
+      Http2ClientMetrics http2ClientMetrics) {
+    this.http2PeerCertificateSanRegex = http2PeerCertificateSanRegex;
+    this.http2ClientMetrics = http2ClientMetrics;
+  }
+
+  @Override
+  public void channelActive(ChannelHandlerContext ctx) throws Exception {
+    logger.trace("Channel Active " + ctx.channel().remoteAddress());
+    validateSslConnection(ctx);
+    super.channelActive(ctx);
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+    logger.trace("Channel Inactive " + ctx.channel().remoteAddress());
+    super.channelInactive(ctx);
+  }
+
+  /**
+   * Listen to the handshake future for the SSL termination, then apply the security policies, potentialy closing the
+   * connection in case security check fails.
+   * @param ctx the {@link ChannelHandlerContext}.
+   */
+  private void validateSslConnection(ChannelHandlerContext ctx) throws Exception {
+    if (http2PeerCertificateSanRegex.trim().isEmpty()) {
+      return;
+    }
+    SslHandler sslHandler = ctx.channel().pipeline().get(SslHandler.class);
+    if (sslHandler == null) {
+      return;
+    }
+    sslHandler.handshakeFuture().addListener(future -> {
+      if (!future.isSuccess()) {
+        logger.error("SSL handshake failed for channel: {}", ctx.channel(), future.cause());
+        return;
+      }
+      try {
+        SSLSession sslSession = sslHandler.engine().getSession();
+        Certificate[] peerCertificates = sslSession.getPeerCertificates();
+        for (Certificate cert : peerCertificates) {
+          if (cert instanceof X509Certificate) {
+            X509Certificate x509Certificate = (X509Certificate) cert;
+            Principal p = x509Certificate.getSubjectDN();
+
+            Collection<?> subjectAltNames = x509Certificate.getSubjectAlternativeNames();
+            if (subjectAltNames == null) {
+              continue;
+            }
+            int numSanName = 0;
+            for (Object subjectAltName : subjectAltNames) {
+              numSanName++;
+              List<?> entry = (List<?>) subjectAltName;
+              if (entry == null || entry.size() < 2) {
+                continue;
+              }
+              Integer altNameType = (Integer) entry.get(0);
```

Comment:
What do you think about also having a config for expected alt name type vs checking all types of alternative names for a match?

Reply:
I feel like that would be an overkill and also error prone due to requiring users to enter a consistent type. Though not very likely, newly added types in future could also cause breakage. If you feel strong about it I can add it though.

Reply:
Nah, I don't feel too strongly about this. Your reasoning makes sense for keeping the configs simple.